### PR TITLE
[5682] Add a UI for bulk publish all

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/5682

### CODE REVIEW

File: ui/app/src/state/reducers/uiConfig.ts [hunk 1]
```diff
@@ -64,6 +64,7 @@ const initialState: GlobalState['uiConfig'] = {
     bulkPublishCommentRequired: false,
     publishByCommitCommentRequired: false,
     publishCommentRequired: false,
+    publishEverythingCommentRequired: false,
```
Comment (rart) [inline ui/app/src/state/reducers/uiConfig.ts:67]: We need to update the blueprints and possibly docs (i.e. the samples and/or default configs).
cc @sumerjabri @alhambrav

File: ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx [hunk 2]
```diff
@@ -253,26 +279,70 @@ export function PublishOnDemandWidget(props: PublishOnDemandWidgetProps) {
...
-  }, [publishStudioFormData, publishGitFormData, mode, bulkPublishCommentRequired, publishByCommitCommentRequired]);
+  }, [
+    publishStudioFormData,
+    publishGitFormData,
+    publishEverythingFormData,
+    mode,
+    bulkPublishCommentRequired,
+    publishByCommitCommentRequired,
+    publishEverythingCommentRequired
+  ]);
```
Comment (rart) [inline ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx]: This looks like it's executed too often and the body of the effect isn't too expensive. Let's ditch the effect and the `useState` for valid and calculate the `isValid` on the fly.

File: ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx [hunk 3]
```diff
@@ -253,26 +279,70 @@ export function PublishOnDemandWidget(props: PublishOnDemandWidgetProps) {
     });
   };
 
+  const onSubmitPublishEverything = () => {
+    console.log('publish everything!', publishEverythingFormData);
```
Comment (rart) [inline ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx]: 👀

File: ui/app/src/services/publishing.ts [hunk 4]
```diff
@@ -137,6 +137,14 @@ export function publishByCommits(
   }).pipe(map(() => true));
 }
 
+export function publishAll(siteId: string, publishingTarget: string, comment: string): Observable<true> {
+  return postJSON('/studio/api/2/publish/all', {
+    siteId,
+    publishingTarget
+    // comment // TODO: pending API update
```
Comment (rart) [inline ui/app/src/services/publishing.ts]: Backend fails if sent? If you can send it anyway best so we don't forget to uncomment and such

File: ui/app/src/components/PublishOnDemandForm/PublishOnDemandForm.tsx [hunk 5]
```diff
@@ -79,46 +79,48 @@ export function PublishOnDemandForm(props: PublishOnDemandFormProps) {
...
-                  ? () => {
-                      setFormData({ path: formData.path.replace(/(.+)(\/$)/, '$1') });
-                    }
-                  : void 0
-              }
-            />
-          </FormControl>
-        </Grid>
-        <Grid item xs={12} md={4}>
+        {(mode === 'studio' || mode === 'git') && (
```
Comment (rart) [inline ui/app/src/components/PublishOnDemandForm/PublishOnDemandForm.tsx]: As long as we don't add more modes, this simplifies a bit
```suggestion
        {mode !== 'all' && (
```

File: ui/app/src/components/PublishOnDemandForm/PublishOnDemandForm.tsx [hunk 6]
```diff
@@ -79,46 +79,48 @@ export function PublishOnDemandForm(props: PublishOnDemandFormProps) {
...
+                    ? () => {
+                        setFormData({ path: formData.path.replace(/(.+)(\/$)/, '$1') });
+                      }
+                    : void 0
+                }
+              />
+            </FormControl>
+          </Grid>
+        )}
+        <Grid item xs={12} md={mode === 'studio' || mode === 'git' ? 4 : 12}>
```
Comment (rart) [inline ui/app/src/components/PublishOnDemandForm/PublishOnDemandForm.tsx]:
```suggestion
        <Grid item xs={12} md={mode === 'all' ? 12 : 4}>
```

File: ui/app/src/services/publishing.ts (hunk 4 above)
Comment (jvega190) [reply ui/app/src/services/publishing.ts]: I tried sending it anyway but it was failing, the API is updated now, so I'll update

File: ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx [hunk 7]
```diff
@@ -133,23 +138,48 @@ export function PublishOnDemandWidget(props: PublishOnDemandWidgetProps) {
...
   const [publishingTargetsError, setPublishingTargetsError] = useState(null);
+  const { bulkPublishCommentRequired, publishByCommitCommentRequired, publishEverythingCommentRequired } = useSelection(
+    (state) => state.uiConfig.publishing
+  );
   const [publishGitFormData, setPublishGitFormData] = useSpreadState<PublishFormData>(initialPublishGitFormData);
-  const [publishGitFormValid, setPublishGitFormValid] = useState(false);
+  const publishGitFormValid =
+    publishGitFormData.commitIds.replace(/\s/g, '') !== '' &&
+    publishGitFormData.publishingTarget !== '' &&
+    (!publishByCommitCommentRequired || !isBlank(publishGitFormData.comment));
```
Comment (rart) [inline ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx]: Cheapest statements first
```suggestion
    !isBlank(publishGitFormData.publishingTarget) &&
    (!publishByCommitCommentRequired || !isBlank(publishGitFormData.comment)) &&
    publishGitFormData.commitIds.replace(/\s/g, '') !== '';
```

File: ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx [hunk 8]
```diff
@@ -133,23 +138,48 @@ export function PublishOnDemandWidget(props: PublishOnDemandWidgetProps) {
...
   const [publishStudioFormData, setPublishStudioFormData] =
     useSpreadState<PublishFormData>(initialPublishStudioFormData);
-  const [publishStudioFormValid, setPublishStudioFormValid] = useState(false);
-  const { bulkPublishCommentRequired, publishByCommitCommentRequired } = useSelection(
-    (state) => state.uiConfig.publishing
+  const publishStudioFormValid =
+    publishStudioFormData.path.replace(/\s/g, '') !== '' &&
+    publishStudioFormData.publishingTarget !== '' &&
+    (!bulkPublishCommentRequired || !isBlank(publishStudioFormData.comment));
```
Comment (rart) [inline ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx]:
```suggestion
    !isBlank(publishStudioFormData.publishingTarget) &&
    (!bulkPublishCommentRequired || !isBlank(publishStudioFormData.comment)) &&
    publishStudioFormData.path.replace(/\s/g, '') !== '';
```

File: ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx [hunk 9]
```diff
@@ -284,10 +328,16 @@ export function PublishOnDemandWidget(props: PublishOnDemandWidgetProps) {
   };
 
   const onSubmitForm = () => {
-    if (mode === 'studio') {
-      onSubmitBulkPublish();
-    } else {
-      onSubmitPublishBy();
+    switch (mode) {
```
Comment (rart) [inline ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx]: Check isValid here too before calling the submit fn

**Review by rart — CHANGES_REQUESTED** [review]

**Review by rart — APPROVED** [review]

File: ui/app/src/state/reducers/uiConfig.ts (hunk 1 above)
Comment (rart) [reply ui/app/src/state/reducers/uiConfig.ts:67]: @jvega190 you can perform bp updates under same ticket

**Comment by jvega190** [comment]
@rart pushed the retrieval of the new config, re-requested the review

**Review by rart — APPROVED** [review]

**Pull request merged**